### DISCUSSION TITLE
Automated generation of transforms with nested Matrix or SuperTable

### DISCUSSION THREAD

**Discussion by dgsiegel** [spacecatninja/craft-imager-x#181] [General]
This is not so much a bug report, but more a short note in case somebody stumbles over the same issue. According to the docs (https://imager-x.spacecat.ninja/usage/generate.html#setting-it-up) and this comment (https://github.com/spacecatninja/craft-imager-x/issues/75#issuecomment-731835193) we know that Imager _does_ support both Matrix and SuperTable. Here's an example starting with a plain entry:

```
<?php

return [  
    'elements' => [
        [
            'elementType' =>  \craft\elements\Entry::class,
            'criteria' => [
                'section' => 'article',
            ],
            'fields' => ['heroImage'],
            'transforms' => ['transform16x9']
        ],       
    ],    
];
```
This will transform a field `heroImage` (which should be an image) in a section called `article` with the transformation `transform16x9`. If `heroImage` would be a field _inside_ a Matrix Block, we would use something like this:

```
[...]
  'fields' => ['blocks:heroImage.image'],
[...]
```

In this example `blocks` is the Matrix field, `heroImage` the type (block type) inside the Matrix, and `image` the actual assets field.

As SuperTables don't have block types, we would have to use this syntax:

```
[...]
  'fields' => ['blocks:*.image'],
[...]
```

As mentioned in https://github.com/spacecatninja/craft-imager-x/issues/75#issuecomment-731835193, Imager doesn't support nested Matrix inside SuperTable or the other way round, however we can use the same element query like this:

```
'elements' => [
  [
    'elementType' =>  \craft\elements\MatrixBlock::class,
    'criteria' => [
      'fieldId' => 1,
      'type' => [
        'heroImage',
        'quote',
        'gallery',
      ],
    ],
    'fields' => [
      'image',
      'imageLeft',
      'imageRight',
      'items:subMatrix.image',
      'items:*.image',
    ],
    'transforms' => ['transform16x9']
  ],
],
```
Unfortunately this gets a little bit uglier, but not complicated. Craft requires the `fieldId` of the block (see https://github.com/craftcms/cms/issues/1789#issuecomment-312103249) which can be found in `/admin/settings/fields`. That will limit your query to this specific Matrix or SuperTable field. Next you can specify all block types you want to transform in `type`. In fields you can specify the individual field name. The trick here is that you can use the same syntax as above (ie. `<matrixFieldHandle>:<matrixTypeHandle>.<fieldHandle>` or `<SuperTableFieldHandle>:*.<fieldHandle>`. Also, these fields will match all fields in the specified types, ie. if `quote` and `heroImage` both have a field `image` you'll catch both with specifying it just once.

This is actually true for all rules you specify: If any rule matches the transformation will be executed. To have two transformations executed for the same field (or a combination of block types and fields) you can either:

```
'elements' => [
  [
    'elementType' =>  \craft\elements\MatrixBlock::class,
    'criteria' => [
      'fieldId' => 1,
      'type' => ['heroImage']
    ],
    'fields' => ['image'],
    'transforms' => ['transform16x9']
  ],
],
'elements' => [
  [
    'elementType' =>  \craft\elements\MatrixBlock::class,
    'criteria' => [
      'fieldId' => 1,
      'type' => ['heroImage']
    ],
    'fields' => ['image'],
    'transforms' => ['transform4x3']
  ],
],
```

or

```
'elements' => [
  [
    'elementType' =>  \craft\elements\MatrixBlock::class,
    'criteria' => [
      'fieldId' => 1,
      'type' => ['heroImage']
    ],
    'fields' => ['image'],
    'transforms' => ['transform16x9', 'transform4x3']
  ],
],
```

That allows you to get creative in how you combine different blocks and fields with their respective transformations.

Hope this helps!

**Comment by aelvan** [comment]
Great summary. Converted it to a discussion to keep it as a resource for other users.

**Comment by ArtDepartmentMJ** [comment]
Does this still appy for Craft 5? I'm struggling to get auto generation of matrix block images (not nested)
